client/src/main/java/org/apache/celeborn/client/write/InFlightTracker.java

```diff
@@ -0,0 +1,152 @@
+    Set<Integer> batchIdSetPerPair =
+        batchIdPerAddressPair.computeIfAbsent(hostAndPushPort, id -> ConcurrentHashMap.newKeySet());
+    batchIdSetPerPair.add(batchId);
+  }
+
+  public void removeFlightBatches(int batchId, String hostAndPushPort) {
+    Set<Integer> batchIdSetPerPair = batchIdPerAddressPair.get(hostAndPushPort);
+    batchIdSetPerPair.remove(batchId);
+    if (batchIdSetPerPair.size() == 0) {
+      batchIdPerAddressPair.remove(hostAndPushPort);
+    }
+  }
+
+  public ConcurrentHashMap<String, Set<Integer>> getBatchIdPerAddressPair() {
+    return batchIdPerAddressPair;
+  }
+
+  public Set<Integer> getBatchIdSetByAddressPair(String addressPair) {
+    return batchIdPerAddressPair.computeIfAbsent(
+        addressPair, pair -> ConcurrentHashMap.newKeySet());
+  }
+
+  public boolean limitMaxInFlight(String hostAndPushPort) throws IOException {
+    if (pushState.exception.get() != null) {
+      throw pushState.exception.get();
+    }
+
+    Set<Integer> batchIdSet = getBatchIdSetByAddressPair(hostAndPushPort);
+    long times = timeoutMs / delta;
+    try {
+      while (times > 0) {
+        if (batchIdSet.size() <= maxInFlight) {
+          break;
+        }
+        if (pushState.exception.get() != null) {
+          throw pushState.exception.get();
+        }
+        Thread.sleep(delta);
+        times--;
+      }
+    } catch (InterruptedException e) {
+      pushState.exception.set(new IOException(e));
+    }
+
+    if (times <= 0) {
+      logger.warn(
+          "After waiting for {} ms, "
+              + "there are still {} batches in flight "
+              + "for hostAndPushPort {}, "
+              + "which exceeds the limit {}.",
+          timeoutMs,
+          batchIdSet.size(),
+          hostAndPushPort,
+          maxInFlight);
+    }
+
+    if (pushState.exception.get() != null) {
+      throw pushState.exception.get();
+    }
+
+    return times <= 0;
+  }
+
+  public boolean limitZeroInFlight() throws IOException {
+    if (pushState.exception.get() != null) {
+      throw pushState.exception.get();
+    }
+    long times = timeoutMs / delta;
+
+    try {
+      while (times > 0) {
+        if (batchIdPerAddressPair.size() == 0) {
+          break;
+        }
+        if (pushState.exception.get() != null) {
+          throw pushState.exception.get();
+        }
+        Thread.sleep(delta);
+        times--;
+      }
+    } catch (InterruptedException e) {
+      pushState.exception.set(new IOException(e));
+    }
+
+    if (times <= 0) {
+      logger.error(
+          "After waiting for {} ms, there are still {} batches in flight, expect 0 batches",
+          timeoutMs,
+          batchIdPerAddressPair.values().stream().map(Set::size).reduce(Integer::sum));
+    }
+
+    if (pushState.exception.get() != null) {
+      throw pushState.exception.get();
+    }
+
+    return times <= 0;
+  }
+
+  protected int nextBatchId() {
+    return batchId.getAndIncrement();
```

Comment:
Origin code use `addAndGet`, here use `getAndIncrement`, seems not consistent.

Reply:
I will update it.